romanisim/nonlinearity.py

```diff
@@ -187,57 +129,55 @@ def correct(counts, coeffs, reversed=False):
 
 
 class NL:
-    """Keep track of non-linearity coefficients.
-
-    This is a wrapper class to help other classes need to know less
-    about non-linearity, but it doesn't presently do much more than
-    cache the non-linearity derivative coefficients so that they may be
-    used multiple times in the up-the-ramp sampling.
+    """Keep track of non-linearity and inverse non-linearity coefficients.
 
-    It would be nice to encapsulate more information about the loss of
-    photons up the read that is presently in `l1.py`, but that's pretty
-    tightly coupled to the apportionment mechanism and so I haven't
-    explored that adequately.
     """
-    def __init__(self, coeffs):
+    def __init__(self, coeffs, gain=1.0, dq=None):
         """Construct an NL class handling non-linearity correction.
 
         Parameters
         ----------
         coeffs : np.ndarray[ncoeff, nx, ny] (float)
             Non-linearity coefficients from reference files.
-        """
-        self.coeffs = repair_coefficients(coeffs)
-        self.derivcoeffs = derivative(self.coeffs)
-        self.coeffs = self.coeffs[::-1, ...].copy()
-        self.derivcoeffs = self.derivcoeffs[::-1, ...].copy()
-
-    def efficiency(self, counts):
-        """Compute the efficiency of photon counting given a count level.
 
-        Parameters
-        ----------
-        counts : np.ndarray[nx, ny] (float)
-            The counts in each pixel
+        gain : float or np.ndarray[float]
+            Gain (electrons / count) for converting counts to electrons
 
-        Returns
-        -------
-        efficiency : np.ndarray[nx, ny] (float)
-            The efficiency of each pixel at the current count level.
+        dq : np.ndarray[n_resultant, nx, ny]
+            Data Quality array
         """
-        return efficiency(counts, self.derivcoeffs, reversed=True)
-
-    def correct(self, counts):
+        # self.coeffs = repair_coefficients(coeffs)
+        self.gain = gain
+        if dq is not None:
+            self.coeffs, self.dq = repair_coefficients(coeffs, dq)
+        else:
+            self.coeffs = repair_coefficients(coeffs)
+            self.dq = None
```

Comment:
Is self.dq ever used?  I guess ideally in make_l1 or something the problematic linearity bits should get or'ed into the image dq bits?

Reply:
Refactored this a bit.